### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by turkycat**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -221,13 +221,15 @@ public MyObject()
 	arduino.AnalogPinUpdateEvent += MyAnalogPinUpdateCallback;
 
 	//subscribe to the ConnectionEstablished event with the name of the function to be called.
-	bt.ConnectionEstablished += MyConnectionEstablishedCallback;
+	connection.ConnectionEstablished += MyConnectionEstablishedCallback;
 	
 	//always begin your IStream object
-	bt.begin( 115200, 0 );
+	connection.begin( 115200, 0 );
```
Comment (zfields) [inline README.md]: SerialConfig.SERIAL_8N1

File: advanced.md [hunk 2]
```diff
@@ -62,13 +62,13 @@ public MyObject()
...
 	//if you create the firmata client yourself, don't forget to begin it!
-	firmata.begin( bt );
+	firmata.begin( connection );
 	
 	//since we do not care about inputs in this example, we will not call firmata.startListening()!
 	//firmata.startListening();
 	
 	//you must always call 'begin' on your IStream object to connect.
-	bt.begin( 115200, 0 );
+	connection.begin( 115200, 0 );
```
Comment (zfields) [inline advanced.md]: same

File: advanced.md [hunk 3]
```diff
@@ -45,7 +45,7 @@ Whenever you construct the `UwpFirmata` object yourself, you **must** call `star
 
 ```c#
 //member variables
-IStream bluetooth;
+IStream connection;
```
Comment (ooeygui) [inline advanced.md]: Might be less confusing to call this bluetoothSerial - minimizes the number of visual concepts and cognitive load for the user (There are 3 - IStream, connection, Bluetooth, Serial)

File: source/Serial/USBSerial.cpp [hunk 4]
```diff
@@ -266,7 +266,7 @@ UsbSerial::write(
     if ( _current_store_operation && _current_store_operation->Status == Windows::Foundation::AsyncStatus::Error )
```
Comment (zfields) [inline source/Serial/USBSerial.cpp]: Is this still necessary?

File: advanced.md (hunk 3 above)
Comment (turkycat) [inline advanced.md]: proposing to change just the variable name? My intention is to illustrate that all of our connection types are IStream descendents

**Comment by zfields** [comment]
@ras0219-msft Will you look through the *Serial.[h|cpp] files? Specifically the daisy-chain error handling.

File: source/Serial/BluetoothSerial.cpp [hunk 5]
```diff
@@ -195,6 +195,30 @@ BluetoothSerial::flush(
...
+    create_task( _current_store_operation )
+        .then( [ this ]( unsigned int value_ )
+    {
+        return _tx->FlushAsync();
+    } )
+        .then( [ this ]( task<bool> task_ )
+    {
+        try
+        {
+            task_.get();
```
Comment (ras0219-msft) [inline source/Serial/BluetoothSerial.cpp]: Note: ignoring the return value from the previous task.

File: source/Serial/USBSerial.cpp [hunk 6]
```diff
@@ -209,6 +209,17 @@ UsbSerial::flush(
     )
 {
     _current_store_operation = _tx->StoreAsync();
+    create_task( _current_store_operation )
+        .then( [this]( task<unsigned int> task_ )
+    {
+        //detect disconnection
```
Comment (ras0219-msft) [inline source/Serial/USBSerial.cpp]: Can StoreAsync() fail? If so, you should be calling task_.get() under a try block to catch the exception. If not, you don't need a task-based continuation here.

File: README.md [hunk 7]
```diff
@@ -134,8 +134,8 @@ This section explains the basic usage of Windows Remote Arduino. This is an exce
         using namespace Microsoft.Maker.Serial;
 
 		//create a bluetooth connection and pass it to the RemoteDevice
-		IStream bt = new BluetoothSerial();
-        RemoteDevice arduino = new RemoteDevice( bt );
+		IStream connection = new BluetoothSerial( "MyBluetoothDevice" );
```
Comment (ooeygui) [inline README.md]: I don't see where in the documentation "MyBluetoothDevice" is described. Why is this required? Also, I noticed that it is not consistent in the documentation. The Advanced readme uses the previous syntax

File: advanced.md (hunk 3 above)
Comment (ooeygui) [inline advanced.md]: Yes, the name of the variable conflicts with the other names. This is not critical, but consider this part of the documentation

File: advanced.md [hunk 8]
```diff
@@ -167,4 +167,112 @@ public void toggleAllDigitalPins( bool setPinsHigh )
 		firmata.sendDigitalPort( 1, 0 );	//all pins low
```
Comment (ooeygui) [inline advanced.md]: Why are these not #defines? (i.e. HIGH, LOW, D1, etc)

File: advanced.md [hunk 9]
```diff
@@ -167,4 +167,112 @@ public void toggleAllDigitalPins( bool setPinsHigh )
...
+	{
+		int pinValue = 0;
+		while( !writerThreadShouldExit )
+		{
+			if( pinValue > 127 ) pinValue = 0;
+
+			//this is our "critical section". Since pin 10 must always be twice the value of pin 9, we're going to lock
+			//before we send these messages. This guarantees that no other thread can use this UwpFirmata object while we 'own' the lock.
+			//it would be very bad if someone tried to read pin 13 in between our two messages!
+			firmata.@lock();
```
Comment (ooeygui) [inline advanced.md]: Why not document and use lock(firmata) { }?

File: source/Firmata/Firmata/Firmata.cpp [hunk 10]
```diff
@@ -26,27 +26,27 @@
 
 void
 FirmataClass::sendValueAsTwo7bitBytes(
-	int value
-	)
+    int value
```
Comment (ooeygui) [inline source/Firmata/Firmata/Firmata.cpp]: Since this file comes from another source, it is likely best to not modify it.

File: source/Firmata/UwpFirmata.cpp [hunk 11]
```diff
@@ -375,28 +318,28 @@ UwpFirmata::unlock(
...
 
 void
 UwpFirmata::stopThreads(
-	void
-	)
+    void
+    )
 {
-	_input_thread_should_exit = true;
+    _input_thread_should_exit = true;
```
Comment (ooeygui) [inline source/Firmata/UwpFirmata.cpp]: Race condition on thread exit. you should wait until the thread fully exits. Consider that this class could be cleaned up before the thread exits - boom.

File: source/Serial/NetworkSerial.cpp [hunk 12]
```diff
@@ -75,132 +75,148 @@ NetworkSerial::~NetworkSerial(
...
+    else if( _current_load_operation->Status != Windows::Foundation::AsyncStatus::Started ) {
+        // Attempt to detect disconnection
+        if( _current_load_operation->Status == Windows::Foundation::AsyncStatus::Error )
+        {
+            _connection_ready = false;
+            ConnectionLost( L"A fatal error has occurred in NetworkSerial::read() and your connection has been lost." );
+            return -1;
+        }
+
+        _current_load_operation = _rx->LoadAsync( 100 );
```
Comment (ooeygui) [inline source/Serial/NetworkSerial.cpp]: This 100 should be a constant with a descriptive name. Comment why it is 100.

File: source/Serial/NetworkSerial.cpp [hunk 13]
```diff
@@ -75,132 +75,148 @@ NetworkSerial::~NetworkSerial(
...
+
+    if( available() ) {
+        c = _rx->ReadByte();
+    }
+    else if( _current_load_operation->Status != Windows::Foundation::AsyncStatus::Started ) {
+        // Attempt to detect disconnection
+        if( _current_load_operation->Status == Windows::Foundation::AsyncStatus::Error )
+        {
+            _connection_ready = false;
+            ConnectionLost( L"A fatal error has occurred in NetworkSerial::read() and your connection has been lost." );
```
Comment (ooeygui) [inline source/Serial/NetworkSerial.cpp]: This string is used lots through the code. Pull into a constant - better than copy paste reuse.

File: source/Serial/NetworkSerial.cpp [hunk 14]
```diff
@@ -75,132 +75,148 @@ NetworkSerial::~NetworkSerial(
...
-	{
-		_connection_ready = false;
-		ConnectionLost();
-		return 0;
-	}
-
-	_tx->WriteByte( c_ );
-	return 1;
+    // Check to see if connection is ready
+    if( !connectionReady() ) { return 0; }
```
Comment (ooeygui) [inline source/Serial/NetworkSerial.cpp]: If the connection is not ready, you are losing data. Should this be queued into a buffer with a regular drain?

File: README.md (hunk 7 above)
Comment (turkycat) [inline README.md]: You're right about using the previous syntax, we did remove the default constructor and this needs to be removed. The "MyBluetoothDevice" represents a device name or id, as either are accepted as the string constructor

File: advanced.md (hunk 9 above)
Comment (turkycat) [inline advanced.md]: The UwpFirmata object uses a mutex internally, rather than locking on itself. Best practice is to continue using the same lock

File: advanced.md (hunk 8 above)
Comment (turkycat) [inline advanced.md]: Without PinConfig implemented, we won't know how many port defines are valid, and the '0' represents the state of 8 pins as one pin per bit.

File: advanced.md (hunk 9 above)
Comment (ooeygui) [inline advanced.md]: Have you ever read Pragmatic Programmer series?

File: source/Firmata/UwpFirmata.cpp (hunk 11 above)
Comment (turkycat) [inline source/Firmata/UwpFirmata.cpp]: The destructor calls finish() which joins this thread after calling stopThreads. The join logic could be moved to the stopThreads() function.

File: source/Serial/NetworkSerial.cpp (hunk 14 above)
Comment (turkycat) [inline source/Serial/NetworkSerial.cpp]: It could be done, but I have avoided complicating simple classes. I feel that the use of this class implies needing a connection, and a 0 is returned if the data is not written.

File: advanced.md (hunk 8 above)
Comment (ooeygui) [inline advanced.md]: You can still use LOW, since you are sending to the port not the pin.

Comment (turkycat) [inline advanced.md]: Since the method parameter is uint8_t, it would require:
static_cast<uint8_t>( Firmata::PinState::LOW )
and I think it misrepresents the purpose of the State enums. There are 256 possible combinations of values to use here, and using LOW only works for one.

Comment (ooeygui) [inline advanced.md]: Magic numbers are ill-advised. Constants disambiguate states, and should be used in favor of a specific interpretation of the value. If your concern is about conflagrating, introduce a new value for ALLLOW.

Comment (turkycat) [inline advanced.md]: I understand the issues with magic numbers, and typically I would agree. I don't agree in this case, as this is just a simple example of advanced usage of the Firmata 'sendPort'. This value '0' as it is currently written is further explained in the comment, is explained in great detail in the section above, and is one of four total usages in the code sample.

Nevertheless, I propose '0' being changed to '0x00' to help illustrate the fact that it is a binary representation of the state of 8 pins, not a magic value.

Comment (ooeygui) [inline advanced.md]: Thank you for the offline explanation. In this case, I see what you were trying to do and approve.
